src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostConfiguration.cs

```diff
@@ -0,0 +1,187 @@
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNetCore.Components.WebAssembly.Hosting
+{
+    /// <summary>
+    /// WebAssemblyHostConfiguration is a class that implements the interface of an IConfiguration,
+    /// IConfigurationRoot, and IConfigurationBuilder. It can be used to simulatneously build
+    /// and read from a configuration object.
+    /// </summary>
+    public class WebAssemblyHostConfiguration : IConfiguration, IConfigurationRoot, IConfigurationBuilder
+    {
+        private readonly List<IConfigurationProvider> _providers = new List<IConfigurationProvider>();
+        private readonly List<IConfigurationSource> _sources = new List<IConfigurationSource>();
+
+        private readonly List<IDisposable> _changeTokenRegistrations = new List<IDisposable>();
+        private ConfigurationReloadToken _changeToken = new ConfigurationReloadToken();
+
+        /// <summary>
+        /// Returns the sources used to obtain configuration values.
+        /// </summary>
+        public IList<IConfigurationSource> Sources => _sources.ToList();
+
+        /// <summary>
+        /// Returns the providers user to obtain configuration values.
+        /// </summary>
+        public IEnumerable<IConfigurationProvider> Providers => _providers.ToList();
+
+        /// <summary>
+        /// Gets a key/value collection that can be used to share data between the <see cref="IConfigurationBuilder"/>
+        /// and the registered <see cref="IConfigurationProvider"/>s.
+        ///
+        /// In this implementation, this largely exists as a way to satisfy the
+        /// requirements of the IConfigurationBuilder and is not populated by
+        /// <see cref="WebAssemblyHostConfiguration"/> with any meaningful info.
+        /// </summary>
+        IDictionary<string, object> IConfigurationBuilder.Properties { get; } = new Dictionary<string, object>();
+
+        /// <inheritdoc />
+        public string this[string key]
+        {
+            get {
+                // Iterate through the providers in reverse to extract
+                // the value from the most recently inserted provider.
+                for (var i = _providers.Count - 1; i >= 0; i--)
+                {
+                    var provider = _providers[i];
+
+                    if (provider.TryGet(key, out var value))
+                    {
+                        return value;
+                    }
+                }
+
+                return null;
+            }
+            set {
+                if (_providers.Count == 0)
+                {
+                    throw new InvalidOperationException("Can only set property if at least one provider has been inserted.");
+                }
+
+                foreach (var provider in _providers)
+                {
+                    provider.Set(key, value);
```

Comment:
I'm not an expert on the configuration APIs, but find this a bit surprising. Do we definitely want to set the value on *all* the providers? Would it suffice to have a single internal "mutable provider" which is where we write things that get set this way? Then the `_providers.Count == 0` check could be dropped too.

Reply:
Do we expect this to be the common case? I somewhat agree with Steve here. I would keep an instance of the last in memory provider and reuse that if possible instead of setting all the providers. What if a provider is "read-only"?

Reply:
Nevermind, this does the same thing as https://github.com/dotnet/runtime/blob/master/src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationRoot.cs#L52-L80

Reply:
Yes, it's doing the same thing as the ConfigurationRoot.

I did some GitHub archaeology and this code [was introduced](https://github.com/aspnet/Configuration/commit/8edcceec69304365d8392fef85ebe68854ef13f0) about five years ago in the configuration provider.

I had assumed that the goal was to ensure that all providers were in sync with each other. So if I set a value using `Configuration["test"]`, I would expect all child providers to set that value.

Also, I think having the case that errors if no providers are registered is helpful. Maintaining an internal `MemoryProvider` seems like it would be surprising to the end-user.

Reply:
OK, thanks for figuring that out!